query/physicalplan/aggregate_concurrent.go

```diff
@@ -0,0 +1,276 @@
+	pool     memory.Allocator
+	buffer   []arrow.Record
+	nextPlan PhysicalPlan
+	aggFunc  AggregationFunction
+	// emitRecordSize is the size of the records of aggregations that will be
+	// passed to the next callback
+	emitRecordSize int
+}
+
+func ConcurrentAggregate(
+	pool memory.Allocator,
+	aggFunc AggregationFunction,
+) *ConcurrentHashAggregate {
+	return &ConcurrentHashAggregate{
+		pool:           pool,
+		buffer:         make([]arrow.Record, 0),
+		aggFunc:        aggFunc,
+		emitRecordSize: 1024, // TODO magic number
+	}
+}
+
+func (a *ConcurrentHashAggregate) SetNextPlan(nextPlan PhysicalPlan) {
+	a.nextPlan = nextPlan
+}
+
+func (a *ConcurrentHashAggregate) Callback(record arrow.Record) error {
+	a.buffer = append(a.buffer, record)
+	return nil
+}
+
+func (a *ConcurrentHashAggregate) Finish() error {
+	combined := a.combineRecords(a.buffer)
+	for _, record := range combined {
+		err := a.nextPlan.Callback(record)
+		if err != nil {
+			return err
+		}
+	}
+
+	return a.nextPlan.Finish()
+}
+
+func (a *ConcurrentHashAggregate) combineRecords(records []arrow.Record) []arrow.Record {
+	schema := getFinalSchema(records)
+	results := make([]arrow.Record, 0)
+
+	// create a list of builders for each column in the data set (based on the
+	// schema of the first record, as all should have the same schema)
+	var numRows int
+	var resultBuilders []array.Builder
+	resetBuilders := func() {
+		numRows = 0
+		resultBuilders = make([]array.Builder, 0)
+		for _, field := range schema.Fields() {
+			resultBuilders = append(resultBuilders, array.NewBuilder(a.pool, field.Type))
+		}
+	}
+	resetBuilders()
+
+	buildRecord := func() arrow.Record {
+		// combine our result builders into the list of columns
+		cols := make([]arrow.Array, 0)
+		for _, builder := range resultBuilders {
+			cols = append(cols, builder.NewArray())
+		}
+		return array.NewRecord(schema, cols, int64(numRows))
+	}
+
+	// traverse the merge tree and for each path (which is a tuple in our result
+	// set), add the values into each of the result builders
+	mergeTree := a.buildMergeTree(records, schema)
+	a.traverseAndAggregate(mergeTree, make([]interface{}, 0), func(pathTuple []interface{}, array2 arrow.Array) {
+		numRows++
+		for i, val := range pathTuple {
+			appendArrayVal(resultBuilders[i], val)
+		}
+
+		// aggregate the results from each parallel exeuction
+		aggArray, _ := a.aggFunc.Aggregate(a.pool, []arrow.Array{array2})
+		aggResultBuilder := resultBuilders[len(resultBuilders)-1]
+		appendArrayVal(aggResultBuilder, getArrayVal(aggArray, 0))
+
+		// when we reach the trheshold for length of the record, build and add to results
+		if numRows >= a.emitRecordSize {
+			results = append(results, buildRecord())
+			resetBuilders()
+		}
+	})
+
+	// build the final record
+	if numRows > 0 {
+		results = append(results, buildRecord())
+	}
+
+	return results
+}
+
+var nilTreeNode = struct{}{}
+
+// buildMergeTree creates a tree where each level of the tree is a column in
```

Comment:
I don't think I really understand what's going on here. Since https://github.com/polarsignals/frostdb/pull/92 we have a predictable arrow schema in every record, so we can treat each record equally in terms of its columns.

I think there might have been a mismatch in communication. When I said we could build a tree to merge things, I meant to merge two whole records with each other at once.

Overall though I think we shouldn't complicate in the first version, and just combine records when the `Callback` is called, instead of when calling `Finish`.